src/atc/transformations.py

```diff
@@ -197,3 +204,168 @@ def join_time_series_dataframes(
+        df = df.filter(" AND ".join(f"({col} is NOT NULL)" for col in join_cols))
+
+    # Load data from the target table for the purpose of incremental load
+    # if not incremental_load:
+    #    return df.write \
+    #        .format(table_format) \
+    #        .mode("overwrite") \
+    #        .save(target_table_name)
+
+    df_target = Spark.get().table(target_table_name)
+
+    # If the target is empty, always do faster full load
+    if len(df_target.take(1)) == 0:
+        return df.write \
+            .format(table_format) \
+            .mode("overwrite") \
+            .save(target_table_name)
+
+    # Find records that need to be updated in the target (happens seldom)
+
+    # Define the column to be used for checking for new rows
+    # Checking the null-ness of one right row is sufficient to mark the row as new, since null keys are disallowed.
+    key_col = join_cols[-1]
+
+    # Write the filtering string including casting maps to strings
+    filter_string = f"{key_col} IS NULL"
+    for col, col_typ in df.dtypes:
+        if col not in join_cols:
+            if col_typ == "map":
+                filter_string += f" OR (CAST(a.{col} AS varchar) <> CAST(b.{col} AS varchar))"
+            else:
+                filter_string += f" OR (a.{col} <> b.{col})"
+                filter_string += f" OR (a.{col} IS NULL AND b.{col} IS NOT NULL)"
+                filter_string += f" OR (a.{col} IS NOT NULL AND b.{col} IS NULL)"
+
+    df = (df.alias("a")
+          .join(df_target.alias("b"), on=join_cols, how="left")
+          .filter(filter_string)
+          .withColumn("is_new",
+                      F.when(f.col(f"{key_col}").isNull(), True)
+                      .otherwise(False))
+          .select("a.*", "is_new")
+          .cache()
+          )
+
+    # If merge is not required, the data can just be appended
+    merge_required = (len(df.filter(~ F.col("is_new")).take(1)) > 0)
+    df = df.drop("is_new")
+
+    if not merge_required:
+        return df.write \
+            .format(table_format) \
+            .mode("append") \
+            .save(target_table_name)
+
+    unique_id = uuid.uuid4().hex
+    temp_view_name = f"source_{unique_id}"
+    df.createOrReplaceTempView(temp_view_name)
+
+    non_join_cols = [col for col in df.columns if col not in join_cols]
+
+    merge_sql_statement = f"""
+        MERGE INTO {target_table_name} AS target
+        USING {temp_view_name} as source
+        ON
+        {" AND ".join(f"(source.{col} = target.{col})" for col in join_cols)}
+        WHEN MATCHED
+        THEN UPDATE -- update existing records
+        SET
+        {', '.join(f"target.{col} = source.{col}" for col in non_join_cols)}
+        WHEN NOT MATCHED
+        THEN INSERT -- insert new records
+        (
+        {', '.join(df.columns)}
+        )
+        VALUES
+        (
+        {', '.join(f"source.{col}" for col in df.columns)}
+        )
```

Comment:
`WHEN MATCHED THEN UPDATE SET * WHEN NOT MATCHED THEN INSERT *` is much more concise and does the same. See the spark documentation.